Call destroy not destroy! on dependent children

naiyt opened the pull request:
This should fix #19761 while still resolving the issue in #12812. This reverts back to calling `destroy` and not `destroy!` on a parent's `has_many dependent: :destroy` children when you call `destroy` on the parent.

In my opinion, this change is important because of the expectation that `destroy` should not raise an exception (which is the case in Rails 3, and in Rails 4 before #13042). This was causing various bugs for us until we discovered that it was because `destroy!` was now getting called on the related objects.

This retains the fix from #13042 by only raising an exception if the necessary objects are not destroyed when doing a replace.

The one difference is that when the replace fails it now raises `ActiveRecord::RecordNotSaved` instead of `ActiveRecord::RecordNotDestroyed`. I think this is a better exception to raise, because I think it's a bit clearer to somebody when they try to save some new objects and they get `RecordNotSaved` instead of `RecordNotDestroyed`. But that might be a breaking change if people are already handling `RecordNotDestroyed` exceptions for those cases, so I could change it back if needed.

egilburg commented on activerecord/test/models/car.rb:
```diff
@@ -24,3 +24,12 @@ class CoolCar < Car
   default_scope { order('name desc') }
 end
+
+class UndestroyableCar < Car
```
How about `IndestructibleCar` :car:

naiyt commented:
@rafaelfranca do you think this is worth revisiting (per your comment on the [original issue](https://github.com/rails/rails/issues/19761))? We worked around it in our code, but I still think something like this would be worth it. If people agree, I can come back to it and fix the merge conflict/clean it up.

arturopie commented:
Why hasn't this fix been merged?

guigs commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+*   Calling `destroy` on an object with `has_many dependent: :destroy` relations
+    will now call `destroy` and not `destroy!` on the related objects. If a 
+    `.replace` call fails due to a `before_destroy`, it will now raise
+    `ActiveRecord::RecordNotDestroyed`.
```
Should be `ActiveRecord::RecordNotSaved`, according to what was implemented.

naiyt committed:
Call destroy not destroy! on dependent children

This is a proposal for a possibly better solution to commit #13042.
Currently, Rails will raise ActiveRecord::RecordNotDestroyed if you call
destroy on a model, but that model has children with dependent: :destroy
set, and those models fail to be destroyed. This causes issues if you
need to perform some action in the parent's before_destroy regardless of
whether the children fail to be destroyed.

This switches back to calling destroy instead of destroy! on the
children. The original issue (#12812) is resolved in this commit by only
raising an ActiveRecord::RecordNotSaved exception when you are replacing
a dependent: :destroy child, and not when calling destroy on the parent.

naiyt commented:
According to #19761 it sounds like multiple people are running into this. Would be great if we could get it reviewed and merged. I went ahead and brought this into line with the most recent Rails master, resolved conflicts, and squashed my commits. Since I've never contributed something to Rails before, can anybody point me to who might be able to review this?

naiyt commented:
I'm gonna close this because it's been open for a year (and I'm kind of tired of seeing it in my list of open PRs). If anybody needs this they should be able to monkey patch it in.
